File: Languages/javascript/algorithms-dsa/patterns/graphs.mjs
/**
 * Graphs Pattern - BFS, Topological Sort, Union-Find, Dijkstra
 *
 * Key Concepts:
 * - BFS: level-order traversal, shortest path in unweighted graphs
 * - Topological Sort: ordering of DAG nodes (prerequisites)
 * - Union-Find: disjoint sets, cycle detection, connected components
 * - Dijkstra: shortest path in weighted graphs (non-negative weights)
 */

// ============================================================
// BFS Shortest Path (Unweighted Graph)
// ============================================================

/**
 * Find shortest path from source to target in an unweighted graph.
 * BFS guarantees shortest path because it explores level by level.
 *
 * @param {Map<number, number[]>} graph - adjacency list
 * @param {number} source
 * @param {number} target
 * @returns {number[]} shortest path, or [] if unreachable
 *
 * Time: O(V + E), Space: O(V)
 */
export function bfsShortestPath(graph, source, target) {
  if (source === target) return [source];

  const visited = new Set([source]);
  const queue = [[source]]; // each entry is a path
  // Alternative: store parent map for O(V) space

  while (queue.length > 0) {
    const path = queue.shift();
    const node = path[path.length - 1];

    const neighbors = graph.get(node) || [];
    for (const neighbor of neighbors) {
      if (visited.has(neighbor)) continue;
      const newPath = [...path, neighbor];

      if (neighbor === target) return newPath;

      visited.add(neighbor);
      queue.push(newPath);
    }
  }

  return []; // unreachable
}

// ============================================================
// Topological Sort - Kahn's Algorithm (BFS-based)
// ============================================================

/**
 * Topological ordering of a DAG using Kahn's algorithm.
 * Start with nodes of in-degree 0, remove them and reduce neighbors' in-degree.
 *
 * @param {number} numNodes - total number of nodes (0-indexed)
 * @param {number[][]} edges - [from, to] directed edges
 * @returns {number[]} topological order, or [] if cycle exists
 *
 * Time: O(V + E), Space: O(V + E)
 */
export function topologicalSort(numNodes, edges) {
  // Build adjacency list and in-degree count
  const adj = Array.from({ length: numNodes }, () => []);
  const inDegree = new Array(numNodes).fill(0);

  for (const [from, to] of edges) {
    adj[from].push(to);
    inDegree[to]++;
  }

  // Start with all nodes that have no prerequisites
  const queue = [];
  for (let i = 0; i < numNodes; i++) {
    if (inDegree[i] === 0) queue.push(i);
  }

  const order = [];
  while (queue.length > 0) {
    const node = queue.shift();
    order.push(node);

    for (const neighbor of adj[node]) {
      inDegree[neighbor]--;
      if (inDegree[neighbor] === 0) {
        queue.push(neighbor);
      }
    }
  }

  // If we couldn't process all nodes, there's a cycle
  return order.length === numNodes ? order : [];
}

// ============================================================
// Union-Find (Disjoint Set Union)
// ============================================================

/**
 * Union-Find with path compression and union by rank.
 * Supports near-constant time union and find operations.
 *
 * Time: O(α(n)) per operation (inverse Ackermann, practically constant)
 */
export class UnionFind {
  constructor(size) {
    this.parent = Array.from({ length: size }, (_, i) => i);
    this.rank = new Array(size).fill(0);
    this.components = size;
  }

  find(x) {
    // Path compression: point directly to root
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }
    return this.parent[x];
  }

  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);

    if (rootX === rootY) return false; // already connected

    // Union by rank: attach smaller tree under larger
    if (this.rank[rootX] < this.rank[rootY]) {
      this.parent[rootX] = rootY;
    } else if (this.rank[rootX] > this.rank[rootY]) {
      this.parent[rootY] = rootX;
    } else {
      this.parent[rootY] = rootX;
      this.rank[rootX]++;
    }

    this.components--;
    return true;
  }

  connected(x, y) {
    return this.find(x) === this.find(y);
  }

  getComponents() {
    return this.components;
  }
}

// ============================================================
// Dijkstra's Algorithm (Shortest Path in Weighted Graph)
// ============================================================

/**
 * Find shortest distances from source to all nodes in a weighted graph.
 * Uses a min-priority queue (simulated with sorted array here).
 *
 * @param {Map<number, [number, number][]>} graph - adjacency list: node -> [[neighbor, weight]]
 * @param {number} source
 * @returns {Map<number, number>} distances from source to each node
 *
 * Time: O((V + E) log V) with proper priority queue
 * Space: O(V + E)
 */
export function dijkstra(graph, source) {
  const dist = new Map();
  // Initialize all distances to Infinity
  for (const node of graph.keys()) {
    dist.set(node, Infinity);
  }
  dist.set(source, 0);

  // Min-heap simulation: [distance, node]
  // Using a simple array + sort (for clarity; use a real heap in production)
  const pq = [[0, source]];
  const visited = new Set();

  while (pq.length > 0) {
    // Extract minimum distance node
    pq.sort((a, b) => a[0] - b[0]);
    const [d, u] = pq.shift();

    if (visited.has(u)) continue;
    visited.add(u);

    const neighbors = graph.get(u) || [];
    for (const [v, weight] of neighbors) {
      const newDist = d + weight;
      if (newDist < dist.get(v)) {
        dist.set(v, newDist);
        pq.push([newDist, v]);
      }
    }
  }

  return dist;
}

// ============================================================
// Tests
// ============================================================

import { test } from "node:test";
import assert from "node:assert/strict";

test("bfsShortestPath - finds shortest path", () => {
  const graph = new Map([
    [1, [2, 3]],
    [2, [4]],
    [3, [4, 5]],
    [4, [5]],
    [5, []],
  ]);
  assert.deepStrictEqual(bfsShortestPath(graph, 1, 5), [1, 3, 5]);
});

test("bfsShortestPath - source equals target", () => {
  const graph = new Map([[1, [2]]]);
  assert.deepStrictEqual(bfsShortestPath(graph, 1, 1), [1]);
});

test("bfsShortestPath - unreachable target", () => {
  const graph = new Map([
    [1, [2]],
    [2, []],
    [3, []],
  ]);
  assert.deepStrictEqual(bfsShortestPath(graph, 1, 3), []);
});

test("topologicalSort - valid DAG", () => {
  // 0 -> 1, 0 -> 2, 1 -> 3, 2 -> 3
  const order = topologicalSort(4, [[0, 1], [0, 2], [1, 3], [2, 3]]);
  // 0 must come before 1, 2; both 1,2 before 3
  assert.strictEqual(order.indexOf(0) < order.indexOf(1), true);
  assert.strictEqual(order.indexOf(0) < order.indexOf(2), true);
  assert.strictEqual(order.indexOf(1) < order.indexOf(3), true);
  assert.strictEqual(order.indexOf(2) < order.indexOf(3), true);
});

test("topologicalSort - cycle detection", () => {
  // 0 -> 1 -> 2 -> 0 (cycle)
  const order = topologicalSort(3, [[0, 1], [1, 2], [2, 0]]);
  assert.deepStrictEqual(order, []);
});

test("topologicalSort - no edges", () => {
  const order = topologicalSort(3, []);
  assert.strictEqual(order.length, 3);
});

test("UnionFind - basic operations", () => {
  const uf = new UnionFind(5);
  assert.strictEqual(uf.connected(0, 1), false);
  uf.union(0, 1);
  assert.strictEqual(uf.connected(0, 1), true);
  uf.union(2, 3);
  uf.union(0, 3);
  assert.strictEqual(uf.connected(1, 2), true);
  assert.strictEqual(uf.getComponents(), 2); // {0,1,2,3} and {4}
});

test("UnionFind - duplicate union returns false", () => {
  const uf = new UnionFind(3);
  assert.strictEqual(uf.union(0, 1), true);
  assert.strictEqual(uf.union(0, 1), false);
});

test("dijkstra - finds shortest distances", () => {
  const graph = new Map([
    [0, [[1, 4], [2, 1]]],
    [1, [[3, 1]]],
    [2, [[1, 2], [3, 5]]],
    [3, []],
  ]);
  const dist = dijkstra(graph, 0);
  assert.strictEqual(dist.get(0), 0);
  assert.strictEqual(dist.get(1), 3); // 0->2->1 (1+2=3) vs 0->1 (4)
  assert.strictEqual(dist.get(2), 1);
  assert.strictEqual(dist.get(3), 4); // 0->2->1->3 (1+2+1=4)
});

test("dijkstra - disconnected node", () => {
  const graph = new Map([
    [0, [[1, 1]]],
    [1, []],
    [2, []],
  ]);
  const dist = dijkstra(graph, 0);
  assert.strictEqual(dist.get(2), Infinity);
});

File: Languages/go/algorithms-dsa/patterns/graphs.go
package main

import "fmt"

/*
  Pattern: Graphs
  Templates: BFS, DFS, Topological Sort, Dijkstra, Union-Find, Cycle Detection
*/

// --- BFS: Shortest path in unweighted graph ---
func bfs(graph map[int][]int, start, end int) []int {
	if start == end {
		return []int{start}
	}
	visited := map[int]bool{start: true}
	queue := []int{start}
	parent := map[int]int{start: -1}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		for _, neighbor := range graph[node] {
			if !visited[neighbor] {
				visited[neighbor] = true
				parent[neighbor] = node
				queue = append(queue, neighbor)
				if neighbor == end {
					return reconstructPath(parent, start, end)
				}
			}
		}
	}
	return nil // no path
}

func reconstructPath(parent map[int]int, start, end int) []int {
	var path []int
	for at := end; at != -1; at = parent[at] {
		path = append(path, at)
	}
	// Reverse
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	return path
}

// --- DFS: Connected components ---
func dfs(graph map[int][]int, node int, visited map[int]bool, component *[]int) {
	visited[node] = true
	*component = append(*component, node)
	for _, neighbor := range graph[node] {
		if !visited[neighbor] {
			dfs(graph, neighbor, visited, component)
		}
	}
}

func connectedComponents(graph map[int][]int, nodes []int) [][]int {
	visited := make(map[int]bool)
	var components [][]int
	for _, node := range nodes {
		if !visited[node] {
			var component []int
			dfs(graph, node, visited, &component)
			components = append(components, component)
		}
	}
	return components
}

// --- Topological Sort (Kahn's Algorithm - BFS-based) ---
func topologicalSort(numCourses int, prerequisites [][]int) ([]int, bool) {
	// Build adjacency list and in-degree count
	graph := make(map[int][]int)
	inDegree := make([]int, numCourses)
	for _, prereq := range prerequisites {
		course, pre := prereq[0], prereq[1]
		graph[pre] = append(graph[pre], course)
		inDegree[course]++
	}

	// Start with nodes having 0 in-degree
	var queue []int
	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	var order []int
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)
		for _, neighbor := range graph[node] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	if len(order) != numCourses {
		return nil, false // cycle detected
	}
	return order, true
}

// --- Union-Find (Disjoint Set Union) ---
type UnionFind struct {
	parent []int
	rank   []int
	count  int // number of components
}

func NewUnionFind(n int) *UnionFind {
	uf := &UnionFind{
		parent: make([]int, n),
		rank:   make([]int, n),
		count:  n,
	}
	for i := range uf.parent {
		uf.parent[i] = i
	}
	return uf
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // path compression
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) bool {
	px, py := uf.Find(x), uf.Find(y)
	if px == py {
		return false // already connected
	}
	// Union by rank
	if uf.rank[px] < uf.rank[py] {
		uf.parent[px] = py
	} else if uf.rank[px] > uf.rank[py] {
		uf.parent[py] = px
	} else {
		uf.parent[py] = px
		uf.rank[px]++
	}
	uf.count--
	return true
}

func (uf *UnionFind) Connected(x, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

// --- Dijkstra: Shortest path in weighted graph ---
type Edge struct {
	To, Weight int
}

func dijkstra(graph map[int][]Edge, start int, n int) []int {
	dist := make([]int, n)
	for i := range dist {
		dist[i] = 1<<31 - 1 // INT_MAX
	}
	dist[start] = 0
	visited := make([]bool, n)

	for i := 0; i < n; i++ {
		// Find unvisited node with minimum distance
		u := -1
		for v := 0; v < n; v++ {
			if !visited[v] && (u == -1 || dist[v] < dist[u]) {
				u = v
			}
		}
		if u == -1 || dist[u] == 1<<31-1 {
			break
		}
		visited[u] = true
		for _, edge := range graph[u] {
			if dist[u]+edge.Weight < dist[edge.To] {
				dist[edge.To] = dist[u] + edge.Weight
			}
		}
	}
	return dist
}

// --- Cycle Detection in Directed Graph (DFS coloring) ---
func hasCycleDirected(graph map[int][]int, n int) bool {
	// 0=white, 1=gray (in stack), 2=black (done)
	color := make([]int, n)
	for i := 0; i < n; i++ {
		if color[i] == 0 {
			if dfsDetectCycle(graph, i, color) {
				return true
			}
		}
	}
	return false
}

func dfsDetectCycle(graph map[int][]int, node int, color []int) bool {
	color[node] = 1 // gray
	for _, neighbor := range graph[node] {
		if color[neighbor] == 1 {
			return true // back edge = cycle
		}
		if color[neighbor] == 0 && dfsDetectCycle(graph, neighbor, color) {
			return true
		}
	}
	color[node] = 2 // black
	return false
}

func main() {
	// Test BFS shortest path
	graph := map[int][]int{
		0: {1, 2},
		1: {0, 3, 4},
		2: {0, 4},
		3: {1, 5},
		4: {1, 2, 5},
		5: {3, 4},
	}
	path := bfs(graph, 0, 5)
	fmt.Println("BFS path 0→5:", path)
	if path[0] == 0 && path[len(path)-1] == 5 && len(path) <= 4 {
		fmt.Println("PASS: BFS shortest path (length:", len(path)-1, "edges)")
	} else {
		panic("FAIL: BFS path")
	}

	// Test Connected Components
	disconnected := map[int][]int{
		0: {1, 2},
		1: {0, 2},
		2: {0, 1},
		3: {4},
		4: {3},
		5: {},
	}
	components := connectedComponents(disconnected, []int{0, 1, 2, 3, 4, 5})
	fmt.Printf("Connected components: %v\n", components)
	if len(components) == 3 {
		fmt.Println("PASS: 3 connected components")
	} else {
		panic(fmt.Sprintf("FAIL: expected 3 components, got %d", len(components)))
	}

	// Test Topological Sort
	// Course 1 requires Course 0, Course 2 requires Course 1, etc.
	prereqs := [][]int{{1, 0}, {2, 1}, {3, 2}, {3, 0}}
	order, valid := topologicalSort(4, prereqs)
	fmt.Printf("Topological order: %v\n", order)
	if valid && len(order) == 4 {
		fmt.Println("PASS: topological sort")
	} else {
		panic("FAIL: topological sort")
	}

	// Test cycle detection in topological sort
	cyclePrereqs := [][]int{{0, 1}, {1, 2}, {2, 0}}
	_, valid = topologicalSort(3, cyclePrereqs)
	if !valid {
		fmt.Println("PASS: cycle detected in prerequisites")
	} else {
		panic("FAIL: should detect cycle")
	}

	// Test Union-Find
	uf := NewUnionFind(7)
	uf.Union(0, 1)
	uf.Union(1, 2)
	uf.Union(3, 4)
	uf.Union(5, 6)
	uf.Union(4, 5)

	if uf.Connected(0, 2) && !uf.Connected(0, 3) && uf.Connected(3, 6) {
		fmt.Println("PASS: Union-Find")
	} else {
		panic("FAIL: Union-Find")
	}
	fmt.Printf("  Components: %d (expected 2)\n", uf.count)
	if uf.count != 2 {
		panic("FAIL: expected 2 components")
	}

	// Test Dijkstra
	weighted := map[int][]Edge{
		0: {{1, 4}, {2, 1}},
		1: {{3, 1}},
		2: {{1, 2}, {3, 5}},
		3: {{4, 3}},
		4: {},
	}
	dist := dijkstra(weighted, 0, 5)
	fmt.Printf("Dijkstra from 0: %v\n", dist)
	if dist[0] == 0 && dist[1] == 3 && dist[3] == 4 && dist[4] == 7 {
		fmt.Println("PASS: Dijkstra shortest paths")
	} else {
		panic("FAIL: Dijkstra")
	}

	// Test directed cycle detection
	cycleGraph := map[int][]int{0: {1}, 1: {2}, 2: {0}}
	if hasCycleDirected(cycleGraph, 3) {
		fmt.Println("PASS: directed cycle detected")
	} else {
		panic("FAIL: should detect directed cycle")
	}

	dagGraph := map[int][]int{0: {1, 2}, 1: {3}, 2: {3}, 3: {}}
	if !hasCycleDirected(dagGraph, 4) {
		fmt.Println("PASS: DAG has no cycle")
	} else {
		panic("FAIL: DAG should not have cycle")
	}
}
